text/1023-splats.rst

```diff
@@ -0,0 +1,340 @@
+                           #     villains: { * },
+                           #     nemesis: { * }
+                           #   }
+    }
+
+* ``[is ...].*`` and ``[is ...].**``: polymorphic variants for the above
+  splat syntaxes.
+
+  An example of ``*``::
+
+    select Person {
+      [is Hero].*  # Expands into
+                   #   {
+                   #      [is Hero].id,
+                   #      [is Hero].name,
+                   #      [is Hero].secret_identity,
+                   #   }
+    }
+
+  An example of ``**``::
+
+    select Person {
+      [is Hero].**  # Expands into
+                    #   {
+                    #      [is Hero].id,
+                    #      [is Hero].name,
+                    #      [is Hero].secret_identity,
+                    #      [is Hero].villains: { * },
+                    #   }
+    }
+
+Splats in free object types
+===========================
+
+This section builds on the concepts introduced in
+`RFC 1022 - Typing free objects & simplifying SDL syntax <./1022-freetypes.rst>`_.
+
+Allowing splats to be used in the EdgeQL's type sub-language (particularly,
+allowing them to be used in free object type declarations) will
+lead to more concise function declarations and type casts.
+
+We propose to extend the free shape type syntax with the following constructs:
+
+* `<type expression>.*`: include all properties from the computed type of
+  ``type expression`` to the final free object's type. Example::
+
+     function validate(data: {
+       Person.*
+     }) -> bool using (...)
+
+     # `data` parameter will accept free objects that have all properties
+     # declared in the Person type (retaining their cardinality bounds & types)
+
+   An example of a more complicated type expression::
+
+     function validate(data: {
+       (Hero | Villain).*,  # will expand into:
+                            #   { required id: uuid, required name: str }
+
+       foo: str,            # add a "foo" property to this free object type
+     }) -> bool using (...)
+
+* `<modifier> <type expression>.*`: include all properties from the computed
+  type expression overriding cardinality.
+
+   An example of including all properties from another type but making
+   them all optional:
+
+     function validate(data: {
+       optional Person.*
+     }) -> bool using (...)
+
+     # `data` parameter will accept free objects that have all properties
+     # declared in the Person type (making them all optional)
+
+  An example of making all expanded fields required:
+
+     function validate(data: {
+       required Hero.*  # will expand into:
+                        #   {
+                        #     required id: uuid,
+                        #     required name: str,
+                        #     required secret_identity: str
+                        #   }
+     }) -> bool using (...)
+
+When we add a ``never`` type eventually (to pave the path to implementing
```

Comment:
Oh, this is not how `never` works usually.

I.e. a thing that contains `never` can never exists. Let me explain it on tuples first.

One way to think of it intuitively is: if you have a function that returns never:
```
function raise_exception() -> never {..}
```
And use that function:
```
SELECT ("test", raise_exception())
```
This tuple is never created, so compiler is free to optimize it out to `SELECT raise_exception()` (this optimization is a bit more complex that it looks like but I'll omit details for brevity).

Now there is a dilemma:
1. Objects don't behave like tuples in our language (with regard to cross-product semantics)
2. But still if you write `SELECT { name := "test", other := raise_exception() }` the free object `{ name: str, other: never }` will never be created and exception raised, so the return type of this `SELECT`-expression is equivalent to `never`.

So unless I'm wrong with semantics of never/exceptions in (2) we should not use that syntax to cherry-pick fields to exclude.

Reply:
> This tuple is never created, so compiler is free to optimize it out to SELECT raise_exception() (this optimization is a bit more complex that it looks like but I'll omit details for brevity).

Hm, that's an interesting observation. FWIW I copied the TypeScript behavior w.r.t. `never` when writing this up. In TypeScript you can use `never` like this specifically in the *type* subset of the language. In the *type* language, `never` behaves differently from *runtime*, only behaving as a hint during the computation of the whole type expression.

That said, I think you're right, in our semantics it would be a bad idea to use `never` the way I suggested in this proposal. Good catch, Paul.

Perhaps a variation of this syntax could work:

```
function foo(param: {
  User.*,  # splat all fields from the User type
  -id,     # exclude the `id` field
}) 
```